app/models/hbx_enrollment.rb

```diff
@@ -1809,12 +1858,35 @@ def notify_of_coverage_start(publish_to_carrier)
     )
   end
 
+  def term_or_expire_enrollment(term_date = nil)
+    if is_shop?
+      enrollment_benefit_application = sponsored_benefit_package.benefit_application
+      if enrollment_benefit_application.terminated? || enrollment_benefit_application.termination_pending?
+        term_or_cancel_enrollment(self, enrollment_benefit_application.end_on, enrollment_benefit_application.termination_reason)
+      elsif enrollment_benefit_application.expired?
+        expire_coverage!
+      end
+    elsif current_year_ivl_coverage? && term_date.present?
+      terminate_coverage!(term_date)
+    elsif  prior_year_ivl_coverage?
+      expire_coverage!
+    end
+  end
+
+  def fetch_reinstatement_date
+    if coverage_expired?
+      is_shop? ? sponsored_benefit_package.end_on.next_day : effective_on.end_of_year.next_day
+    else
+      terminated_on.next_day
+    end
+  end
+
   def reinstate(edi: false)
     return false unless can_be_reinstated?
-    return false if has_active_or_term_exists_for_reinstated_date?
-    reinstate_enrollment = Enrollments::Replicator::Reinstatement.new(self, terminated_on.next_day).build
+    return false if has_active_term_or_expired_exists_for_reinstated_date?
+    reinstate_enrollment = Enrollments::Replicator::Reinstatement.new(self, fetch_reinstatement_date).build
 
-    if self.is_shop?
+    if self.is_shop? && !prior_plan_year_coverage?
```

Comment:
Did not understand why we are not considering `prior_plan_year_coverage` enrollment

Reply:
We should not be update census employee eligibility, so that is the reason we have added this check
```
census_employee = benefit_group_assignment.census_employee
      census_employee.reinstate_employment if census_employee.can_be_reinstated?
```